river/rules/base.py

```diff
@@ -0,0 +1,275 @@
+        else:
+            return f"{self.on} ≠ {self.at}"
+
+
+class HoeffdingRule(base.Estimator, metaclass=abc.ABCMeta):
+    def __init__(
+        self,
+        template_splitter: tree.splitter.base_splitter.Splitter,
+        split_criterion: tree.split_criterion.base_split_criterion.SplitCriterion,
+        **attributes,
+    ):
+        self.template_splitter = template_splitter
+        self.split_criterion = split_criterion
+        self.literals = []
+        self.splitters = {}
+
+        self._total_weight = 0
+        self._last_expansion_attempt_at = 0
+
+        self.nominal_features = set()
+
+        self.__dict__.update(attributes)
+
+    def _hoeffding_bound(self, r_heur, delta):
+        r"""Compute the Hoeffding bound, used to decide how many samples are necessary to expand
+        a decision rule.
+
+        Notes
+        -----
+        The Hoeffding bound is defined as:
+
+        $\\epsilon = \\sqrt{\\frac{R^2\\ln(1/\\delta))}{2n}}$
+
+        where:
+
+        $\\epsilon$: Hoeffding bound.
+        $R$: Range of a random variable.
+        $\\delta$: significance level.
+        $n$: Number of samples.
+
+        Parameters
+        ----------
+        r_heur
+            Range of the split heuristic.
+        delta
+            The significance level.
+        """
+        return math.sqrt(
+            (r_heur * r_heur * math.log(1.0 / delta)) / (2.0 * self.total_weight)
+        )
+
+    @property
+    @abc.abstractmethod
+    def statistics(self):
+        pass
+
+    @statistics.setter
+    @abc.abstractmethod
+    def statistics(self, target_stats):
+        pass
+
+    @property
+    def total_weight(self):
+        return self._total_weight
+
+    @property
+    def last_expansion_attempt_at(self):
+        return self._last_expansion_attempt_at
+
+    def expand(self, delta, tau):
+        """
+        Attempt to expand a decision rule.
+
+        If the expansion succeeds, a new rule with all attributes reset and updated literals is
+        returned along a boolean flag set to `True` to indicate the expansion. If the expansion
+        fails, `self` is returned along with the boolean flag set to `False`.
+
+        Parameters
+        ----------
+        delta
+            The split test significance.
+        tau
+            The tie-breaking threshold.
+
+        Returns
+        -------
+        rule, expanded
+            The (potentially expanded) rule and an indicator of whether or not an expansion was
+            successfully performed.
+
+        """
+        suggestions = []
+        for att_id, splitter in self.splitters.items():
+            suggestions.append(
+                splitter.best_evaluated_split_suggestion(
+                    criterion=self.split_criterion,
+                    pre_split_dist=self.statistics,
+                    att_idx=att_id,
+                    binary_only=True,
+                )
+            )
+        suggestions.sort()
```

Comment:
What is the criterium for this sorting?

Reply:
The split merit. The `BranchFactory` can be sorted :)

https://github.com/online-ml/river/blob/eda2723b10ae22738bdc063fca3846e4ff601597/river/tree/utils.py#L99-L103